### PR TITLE
Shape union

### PR DESCRIPTION
**Pull request by KaiSzuttor**
A union of shapes.

### CODE REVIEW

File: src/core/shapes/ShapeUnion.hpp [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+  }
+  void calculate_dist(Utils::Vector3d const &pos, double &dist,
+                      Utils::Vector3d &vec) const {
+    auto dist_compare = [&pos](std::pair<double, Utils::Vector3d> const &res,
+                               std::shared_ptr<Shapes::Shape> const &s) {
+      double d;
+      Utils::Vector3d vec;
+      (*s).calculate_dist(pos, d, vec);
+      if (d < 0.0)
+        throw std::runtime_error(
```
Comment (fweik) [inline src/core/shapes/ShapeUnion.hpp]: This makes the union unusable as an LB boundary.

**Review by jngrad — COMMENTED** [review]
Looks good, but it would be nice to have a minimal documentation for this feature. For example, the `calculate_dist()` method could use a docstring like `/** Calculate the distance to the closest shape */`. In the Sphinx docs, file `constraints.rst`, one could make section 11.2 a subsection of section 11.1 (and make 11.2.1 to 11.2.4 subsubsections of 11.2, but keep 11.2.5-11.2.6 as subsections of 11.1), and have a new section 11.3 to describe the union.

**Comment by fweik** [comment]
@jngrad this should have a formal definition of what is calculated (which is not hard), but first this has to be actually feature complete ^^

**Comment by jngrad** [comment]
@fweik of course. I saw your comment when writing my review. I was a little confused at first as to what was calculated, and thought the union was a variant, until I read through the boost accumulator code ^^

File: src/python/espressomd/shapes.py [hunk 2]
```diff
@@ -266,3 +269,73 @@ class HollowConicalFrustum(Shape, ScriptInterfaceHelper):
     .. image:: figures/conical_frustum.png
     """
     _so_name = "Shapes::HollowConicalFrustum"
+
+
+@script_interface_register
+class ShapeUnion(Shape, ScriptObjectRegistry):
```
Comment (fweik) [inline src/python/espressomd/shapes.py]: I think this should be called `Union`, it's clear from the context that this is about `Shapes`.

**Review by fweik — APPROVED** [review]
LGTM. The change on the bondary logic need to be applied to #2701 (Walberla).

**Comment by KaiSzuttor** [comment]
@fweik I've simplified the LB boundary initialization a bit via boost.

**Pull request merged**